src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py

```diff
@@ -218,8 +206,8 @@ def has_pending_clones(self, snapname):
             raise
 
     def remove_snapshot(self, snapname):
-        if self.is_snapshot_protected(snapname):
-            raise VolumeException(-errno.EINVAL, "snapshot '{0}' is protected".format(snapname))
+        if self.has_pending_clones(snapname):
```

Comment:
This also protects the snapshot with subvolumes version 0 (i.e., manila shares), which has pending clones?

Maybe later, not necessarily in this PR, we can add a test for this.

Reply:
> This also implicitly protects the snapshot with subvolumes version 0 (i.e., manila shares), which has pending clones?

As the implementation stands now, legacy volumes are [auto upgraded](https://github.com/ceph/ceph/blob/0e76c3a743f48167b56de1f9e5b428a0a16f9532/src/pybind/mgr/volumes/fs/operations/versions/__init__.py#L47-L74), once that is complete all volumes are v1 volumes from a version and code standpoint, there is no further special handling for legacy volumes.

The method to `remove_snapshot` hence remains the same across both versions. IOW, legacy volumes will also pass through the above code to remove a snapshot.

Prior to removing a snapshot, the volume has to have been upgraded to v1 to perform a clone operation on it, which will save the required clone in progress record for the source snapshot, that is checked at this point in the code.

> Maybe later, not necessarily in this PR, we can add a test for this.

Testing for legacy volumes has the downside that these are auto upgraded, hence on first access these would be upgraded to v1 volumes and a test case for the same is present in `test_subvolume_upgrade`.

Now, we can add a test case similar to `test_subvolume_upgrade` where we start with a legacy volume and a snapshot of the same still existing in the legacy format, but the first operation would be `subvolume snapshot clone ...` which will upgrade the volume to v1, and hence the test for `subvolume snapshot rm ...` will already be operating on a v1 volume.

IOW, I am not sure how to add a test case targeting this condition check and failing for legacy subvolumes :)

Reply:
> IOW, I am not sure how to add a test case targeting this condition check and failing for legacy subvolumes :)

Thinking further, a legacy upgraded volume can still test for the clone in progress record maintenance, when the v1 meta file is in the legacy location. So a short test case to ensure sanity makes sense. I will add one with this PR itself.

Reply:
Cloning from legacy subvolume is in test `test_subvolume_snapshot_clone_with_upgrade`. Is that what you are thinking to add as a test?

Reply:
> > This also implicitly protects the snapshot with subvolumes version 0 (i.e., manila shares), which has pending clones?
> 
> As the implementation stands now, legacy volumes are [auto upgraded](https://github.com/ceph/ceph/blob/0e76c3a743f48167b56de1f9e5b428a0a16f9532/src/pybind/mgr/volumes/fs/operations/versions/__init__.py#L47-L74), 

Auto-upgrade is finding the right location of the meta file to modify, right? For subvolume version 0, the subvolume meta file's parent is the directory with subvolumegroup name;  and for subvolume version 1, the subvolume meta file's parent is the directory with the subvolume name.

> once that is complete all volumes are v1 volumes from a version and code standpoint, there is no further special handling for legacy volumes.
> 
> The method to `remove_snapshot` hence remains the same across both versions. IOW, legacy volumes will also pass through the above code to remove a snapshot.
> 

Answer is yes then. Got it.

> Prior to removing a snapshot, the volume has to have been upgraded to v1 to perform a clone operation on it, which will save the required clone in progress record for the source snapshot, that is checked at this point in the code.
> 

> > Maybe later, not necessarily in this PR, we can add a test for this.
> 
> Testing for legacy volumes has the downside that these are auto upgraded, hence on first access these would be upgraded to v1 volumes and a test case for the same is present in `test_subvolume_upgrade`.
> 
> Now, we can add a test case similar to `test_subvolume_upgrade` where we start with a legacy volume and a snapshot of the same still existing in the legacy format, but the first operation would be `subvolume snapshot clone ...` which will upgrade the volume to v1, and hence the test for `subvolume snapshot rm ...` will already be operating on a v1 volume.
> 
> IOW, I am not sure how to add a test case targeting this condition check and failing for legacy subvolumes :)

Reply:
> > > This also implicitly protects the snapshot with subvolumes version 0 (i.e., manila shares), which has pending clones?
> > 
> > 
> > As the implementation stands now, legacy volumes are [auto upgraded](https://github.com/ceph/ceph/blob/0e76c3a743f48167b56de1f9e5b428a0a16f9532/src/pybind/mgr/volumes/fs/operations/versions/__init__.py#L47-L74),
> 
> Auto-upgrade is finding the right location of the meta file to modify, right? For subvolume version 0, the subvolume meta file's parent is the directory with subvolumegroup name; and for subvolume version 1, the subvolume meta file's parent is the directory with the subvolume name.

(short clarification)
- Auto-upgrade (`upgrade_legacy_subvolume`) is creating the meta file in the right location when not found, when legacy/v0 subvolume exists.
- `discover` handles opening the file from the correct location, irrespective of v0/1 subvolumes.

Both are invoked from `get_subvolume_object`.

Reply:
> > > > This also implicitly protects the snapshot with subvolumes version 0 (i.e., manila shares), which has pending clones?
> > > 
> > > 
> > > As the implementation stands now, legacy volumes are [auto upgraded](https://github.com/ceph/ceph/blob/0e76c3a743f48167b56de1f9e5b428a0a16f9532/src/pybind/mgr/volumes/fs/operations/versions/__init__.py#L47-L74),
> > 
> > 
> > Auto-upgrade is finding the right location of the meta file to modify, right? For subvolume version 0, the subvolume meta file's parent is the directory with subvolumegroup name; and for subvolume version 1, the subvolume meta file's parent is the directory with the subvolume name.
> 
> (short clarification)
> 
> * Auto-upgrade (`upgrade_legacy_subvolume`) is creating the meta file in the right location when not found, when legacy/v0 subvolume exists.
> * `discover` handles opening the file from the correct location, irrespective of v0/1 subvolumes.
> 
> Both are invoked from `get_subvolume_object`.

I was sloppy with the details. Got it.

Reply:
> > Cloning from legacy subvolume is in test `test_subvolume_snapshot_clone_with_upgrade`. Is that what you are thinking to add as a test?
> 
> Yes, so this is already covered.

This doesn't cover what I pointed out where we try to delete a legacy subvolume's snapshot when it's being cloned.

Reply:
> This doesn't cover what I pointed out where we try to delete a legacy subvolume's snapshot when it's being cloned.

When a legacy subvolume snapshot is deleted when clone is in progress but the subvolume is not upgraded?

Reply:
> I was sloppy with the details. Got it.

No worries, I was attempting to be more precise as I am new to the code base. FWIW, post your comment I had to read this code a bit more in detail to ensure it does not overwrite the legacy file :)

Reply:
> > This doesn't cover what I pointed out where we try to delete a legacy subvolume's snapshot when it's being cloned.
> 
> When a legacy subvolume snapshot is deleted when clone is in progress but the subvolume is not upgraded?

I think amending the existing test case to delete a snapshot, when clone is in flight, for a legacy volume, helps. Will update the PR with the same.

Reply:
@ajarr @vshankar Added a snapshot remove test when upgrading legacy volumes and cloning the same.